### PR TITLE
ci: add annotation api

### PR DESCRIPTION
**Pull request by asdine**
This adds a script that allows buildkite steps to easily add a annotation when an error occurs.

Related to #26071 

Co-authored-by: Jean Hadrien Chabran <jh@chabran.fr>

### CODE REVIEW

File: dev/check/shellcheck.sh [hunk 1]
```diff
@@ -2,14 +2,19 @@
 
 echo "--- shellcheck"
 
-trap "echo ^^^ +++" ERR
-
-set -e
-
-cd "$(dirname "${BASH_SOURCE[0]}")"/../..
+cd "$(dirname "${BASH_SOURCE[0]}")"/../.. || exit
```
Comment (jhchabran) [inline dev/check/shellcheck.sh:5]: The check would fail silently otherwise if it could not cd to that folder (as we removed the `trap`)

```suggestion
cd "$(dirname "${BASH_SOURCE[0]}")"/../.. || exit 1
```

**Review by jhchabran — APPROVED** [review]

**Review by jhchabran — CHANGES_REQUESTED** [review]
Really small thing that we missed! 

See my comment https://github.com/sourcegraph/sourcegraph/pull/27897/files#r752263483

File: dev/ci/annotate.sh [hunk 2]
```diff
@@ -0,0 +1,55 @@
+#!/bin/bash
+
+cd "$(dirname "${BASH_SOURCE[0]}")/../../"
+set -ex
```
Comment (jhchabran) [inline dev/ci/annotate.sh:4]: @asdine we forgot to remove that `x`, it's going to be noisy otherwise. 
```suggestion
set -e
```

File: dev/check/shfmt.sh [hunk 3]
```diff
@@ -2,10 +2,15 @@
 
 echo "--- shfmt (ensure shell-scripts are formatted consistently)"
 
-trap "echo ^^^ +++" ERR
+cd "$(dirname "${BASH_SOURCE[0]}")"/../.. || exit
 
-set -ex
+OUT=$(shfmt -d .)
+EXIT_CODE=$?
```
Comment (jhchabran) [inline dev/check/shfmt.sh:8]:
```suggestion
set +e
OUT=$(shfmt -d .)
EXIT_CODE=$?
set -e
```

File: dev/check/shfmt.sh [hunk 4]
```diff
@@ -2,10 +2,15 @@
 
 echo "--- shfmt (ensure shell-scripts are formatted consistently)"
 
-trap "echo ^^^ +++" ERR
+cd "$(dirname "${BASH_SOURCE[0]}")"/../.. || exit
```
Comment (jhchabran) [inline dev/check/shfmt.sh:5]:
```suggestion
set -e
cd "$(dirname "${BASH_SOURCE[0]}")"/../..
```

**Review by jhchabran — APPROVED** [review]
💪:shell: 

See the comments, Approving to not have to re-review again!

**Review by bobheadxi — COMMENTED** [review]
This seems to go a bit against the `sg` mantra of _removing_ bash scripts, rather than adding more

I also envisioned that this would be in the generator rather than something that has to be explicitly called like this, which feels a bit complicated

**Comment by jhchabran** [comment]
> This seems to go a bit against the `sg` mantra of _removing_ bash scripts, rather than adding more

Agreed. It's more of a prototype than anything else though, if that proves to be useful, it makes sense to get it back in `sg` or in the pipeline generator. 

> I also envisioned that this would be in the generator rather than something that has to be explicitly called like this, which feels a bit complicated

Yeah, it presently is a bit complicated, which is part of the discovery work on this. The original idea was to simply write the logs into a file, that is then read to create the annotation. The problem with that it's not working when you have concurrent commands being run. 

To explain our reasoning: 

Basically, the problem is as following, either we make the log output better, which is a lot of work, and would potentially make it harder to debug what happened, or we selectively annotate content that we know for sure is relevant to the user. 

There is a lot of value with the former, but the latter yields immediate feedback on what we could be doing with those annotations.

**Comment by bobheadxi** [comment]
> The original idea was to simply write the logs into a file, that is then read to create the annotation. The problem with that it's not working when you have concurrent commands being run.

Allowing an entire directory for annotation output might be a workaround here, where different parts of the command can output things to different files which could then get selectively pulled out to annotations
